internal/expr/bindinputs.go

```diff
@@ -52,33 +40,108 @@ func (te *TypedExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
 			return nil, fmt.Errorf("type %q provided more than once", t.Name())
 		}
 		typeToValue[t] = v
-		if !inQuery[t] {
-			// Check if we have a type with the same name from a different package.
-			for _, input := range te.inputs {
-				if t.Name() == input.ArgType().Name() {
-					return nil, fmt.Errorf("parameter with type %q missing, have type with same name: %q", input.ArgType().String(), t.String())
+	}
+
+	// Generate SQL and query parameters.
+	params := []any{}
+	outputs := []typeinfo.Output{}
+	// argTypeUsed is used to check that all the query parameters are
+	// referenced in the query.
+	argTypeUsed := map[reflect.Type]bool{}
+	inputCount := 0
+	outputCount := 0
+	sqlStr := bytes.Buffer{}
+	for _, te := range *tbe {
+		switch te := te.(type) {
+		case *typedInputExpr:
+			vals, err := te.input.LocateParams(typeToValue)
+			if err != nil {
+				return nil, err
+			}
+			argTypeUsed[te.input.ArgType()] = true
+			for _, val := range vals {
+				namedInput := sql.Named("sqlair_"+strconv.Itoa(inputCount), val.Interface())
+				params = append(params, namedInput)
+				sqlStr.WriteString("@sqlair_" + strconv.Itoa(inputCount))
+				inputCount++
+			}
+		case *typedOutputExpr:
+			for i, oc := range te.outputColumns {
+				sqlStr.WriteString(oc.sql(outputCount))
+				if i != len(te.outputColumns)-1 {
+					sqlStr.WriteString(", ")
 				}
+				outputCount++
+				outputs = append(outputs, oc.output)
 			}
-			return nil, fmt.Errorf("%s not referenced in query", t.Name())
+		case *bypass:
```

Comment:
Given that `typedExpression` is a completely synthetic interface, just delete it.
- `TypeBoundExpr` becomes `[]any`. 
- Add a default case here that returns an error for types we don't recognise.
- Delete all the marker methods and implementation checks below.

If we come up with a unifying indirection/abstraction, we can reintroduce an interface. As it is, it's just noise.

Reply:
I agree. Just want to point out that the only benefit for having a synthetic interface is that it documents which types are allowed and which are not, and it is in code instead of a comment which might get outdated.